### PR TITLE
EOS:16672 update s3confstore APIs

### PR DESCRIPTION
**Pull request by amitwac2608**
Added new APIs for getting privateip, nodescount and hostnames.
Re-factored the s3confstore to be called efficiently from python code.

Signed-off-by: Amit Kumar <amit.kumar@seagate.com>

### CODE REVIEW

File: s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py [hunk 1]
```diff
@@ -26,138 +29,166 @@
...
-      exit(-1)
-    else:
-      Conf.load(index, backend_url)
+    if config is not None:
+      self.validate_configfile(config)
+      self.load_config(self.default_index, self.config_file)
+
+  @staticmethod
+  def load_config(index: str = "default_s3confstore_index", config: str = ""):
+    """Load Config into confstore."""
```
Comment (shahidSTX) [inline s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py:44]: I see that you've removed the check :
_if not backend_url.strip():_
Is this on purpose?

File: s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py [hunk 2]
```diff
@@ -26,138 +29,166 @@
...
+    if os.path.isfile(urlparse(configfile).path) != True:
+      print("config file: {} does not exist".format(configfile))
+      sys.exit(1)
+    else:
+      store_type = urlparse(configfile).scheme
+      is_valid_type = False
+      valid_types = ''
+      storage = inspect.getmembers(kv_store_collection, inspect.isclass)
+      for name, cls in storage:
+        if hasattr(cls, 'name') and name != "KvStore":
```
Comment (amitwac2608) [inline s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py:70]: see if we need to do case-sensitive comparison.

File: s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py [hunk 3]
```diff
@@ -26,138 +29,166 @@
...
+          print("config file: {} must use valid JSON format: {}".format(urlparse(configfile).path, e))
+          sys.exit(1)
+      """TODO: Implement rest of the type's content validators here"""
+
+  def get_nodecount(self):
+    """Get total nodes count in the cluster, from py-utils::confstore."""
+    nodes_count = None
+    key_to_read_from_conf = 'cluster>server_nodes'
+
+    dict_servernodes = self.get_config(key_to_read_from_conf)
```
Comment (amitwac2608) [inline s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py:94]: set dict_servernodes explicitly to None.

File: s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py [hunk 4]
```diff
@@ -26,138 +29,166 @@
...
+          with open(urlparse(configfile).path) as f:
+            json.load(f)
+        except ValueError as e:
+          print("config file: {} must use valid JSON format: {}".format(urlparse(configfile).path, e))
+          sys.exit(1)
+      """TODO: Implement rest of the type's content validators here"""
+
+  def get_nodecount(self):
+    """Get total nodes count in the cluster, from py-utils::confstore."""
+    nodes_count = None
```
Comment (amitwac2608) [inline s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py:91]: set the initialize value as 0

File: s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py [hunk 5]
```diff
@@ -26,138 +29,166 @@
...
+  def get_privateip(self, machine_id: str):
+    """Get private_ip of the host, whose machineid has been passed, from py-utils::confstore."""
+    privateip = ""
+    key_to_read_from_conf = 'cluster>server_nodes'
+
+    dict_servernodes = self.get_config(key_to_read_from_conf)
+
+    # find the 'machine_id' in the keys of dict_servernodes
+    server_node = ""
+    if machine_id in dict_servernodes.keys():
```
Comment (amitwac2608) [inline s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py:111]: None check for dict_servernodes

**Comment by pratyush-seagate** [comment]
Please publish the jenkins and unit test results. Overall changes looks good to me.

File: s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py [hunk 6]
```diff
@@ -26,138 +29,166 @@
...
+                        help='config file url, check cortx-py-utils::confstore for supported formats.',
+                        type=str)
+
+    subparsers = parser.add_subparsers(dest='command', title='comamnds')
+
+    getkey = subparsers.add_parser('getkey', help='get value of given key from confstore')
+    getkey.add_argument('--key', help='Fetch value of the given key', type=str, required=True)
+
+    setkey = subparsers.add_parser('setkey', help='set given value to given key in confstore')
+    setkey.add_argument('--key', help='set new value to given key', type=str)
```
Comment (amitwac2608) [inline s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py:149]: required = True

**Review by amitwac2608 — COMMENTED** [review]
Thanks @sachitanands  for review. I will incorporate the comments.

File: s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py (hunk 2 above)
Comment (amitwac2608) [reply s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py:70]: @pratyush-seagate  - can you respond here ?

File: s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py (hunk 1 above)
Comment (amitwac2608) [reply s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py:44]: I will add this check in validate function. Thanks for pointing this out

File: s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py (hunk 2 above)
Comment (pratyush-seagate) [reply s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py:70]: This code checks the parsing capabilities provided by lower layers. KvStore needs to be avoided in this check. I don't think case-based checks are needed unless that class name changes.
Also this code is existing and works as expected.

File: s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py (hunk 4 above)
Comment (amitwac2608) [reply s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py]: done

File: s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py (hunk 6 above)
Comment (amitwac2608) [reply s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py]: done

File: s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py (hunk 5 above)
Comment (amitwac2608) [reply s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py]: done

File: s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py (hunk 3 above)
Comment (amitwac2608) [reply s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py:100]: done

File: s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py (hunk 1 above)
Comment (amitwac2608) [reply s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py:44]: done

**Review by shahidSTX — APPROVED** [review]
LGTM

**Comment by amitwac2608** [comment]
Jenkins Job passed : http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server-7.8.2003/437/

**Review by pratyush-seagate — APPROVED** [review]

**Pull request merged**